src/DeferredFutures.jl

```diff
@@ -1,35 +1,88 @@
 module DeferredFutures
 
-export DeferredFuture
+export @defer, DeferredChannel, DeferredFuture
 
 using AutoHashEquals
 
-@auto_hash_equals immutable DeferredFuture <: Base.AbstractRemoteRef
+abstract DeferredRemoteRef <: Base.AbstractRemoteRef
+
+@auto_hash_equals immutable DeferredFuture <: DeferredRemoteRef
+    outer::Future
+end
+
+DeferredFuture(pid::Integer=myid()) = DeferredFuture(Future(pid))
+
+@auto_hash_equals immutable DeferredChannel <: DeferredRemoteRef
     outer::Future
+    func::Function      # A function the generates the RemoteChannel
+end
+
+function DeferredChannel(f::Function, pid::Integer=myid())
+    DeferredChannel(Future(pid), ()->RemoteChannel(f, pid))
+end
+
+function DeferredChannel(pid::Integer=myid(), num::Integer=32; content::DataType=Any)
+    DeferredChannel(
+        Future(pid),
+        ()->RemoteChannel(
+            ()->Channel{content}(num)
+        )
+    )
 end
 
-DeferredFuture() = DeferredFuture(Future())
-DeferredFuture(pid::Integer) = DeferredFuture(Future(pid))
 
-function Base.put!(df::DeferredFuture, val)
-    rc = RemoteChannel()
-    put!(rc, val)
-    put!(df.outer, rc)
-    return df
+function Base.put!(ref::DeferredFuture, val)
+    inner = RemoteChannel()
+    put!(ref.outer, inner)
+    put!(fetch(ref.outer), val)
+
+    return ref
+end
+
+function Base.put!(ref::DeferredChannel, val)
+    # On the first call to put! create the `RemoteChannel`
+    # and `put!`` it in the `Future`
+    if !isready(ref.outer)
+        inner = ref.func()
+        put!(ref.outer, inner)
+    end
+
+    # Feature the `RemoteChannel` and `put!`
+    # the value in there
+    put!(fetch(ref.outer), val)
+
+    return ref
 end
 
-function Base.isready(df::DeferredFuture)
-    isready(df.outer) && isready(fetch(df.outer))
+function Base.isready(ref::DeferredRemoteRef)
+    isready(ref.outer) && isready(fetch(ref.outer))
 end
 
-function Base.fetch(df::DeferredFuture)
-    fetch(fetch(df.outer))
+function Base.fetch(ref::DeferredRemoteRef)
+    fetch(fetch(ref.outer))
 end
 
-function Base.wait(df::DeferredFuture)
-    wait(df.outer)
-    wait(fetch(df.outer))
-    return df
+function Base.wait(ref::DeferredRemoteRef)
+    wait(ref.outer)
+    wait(fetch(ref.outer))
+    return ref
+end
+
+Base.getindex(ref::DeferredRemoteRef) = getindex(fetch(ref.outer))
+Base.close(ref::DeferredChannel) = close(fetch(ref.outer))
+Base.take!(ref::DeferredChannel) = take!(fetch(ref.outer))
+
+macro defer(ex::Expr)
+    @assert ex.head == :call
+    @assert isa(ex.args[1], Symbol)
+
+    if ex.args[1] == :Future
+        return Expr(:call, :DeferredFuture, ex.args[2:end]...)
+    elseif ex.args[1] == :RemoteChannel
+        return Expr(:call, :DeferredChannel, ex.args[2:end]...)
+    else
+        return parse("throw(AssertionError(\"Expected RemoteChannel or Future and got $(ex.args[1])\"))")
```

Comment:
I disagree because this makes testing that it throws the AssertionError difficult. A try catch doesn't seem to catch the AssertionError if it is thrown inside the macro.